File: app/src/main/java/com/princeparadoxes/watertracker/data/repository/DrinkRepository.java
package com.princeparadoxes.watertracker.data.repository;

import com.princeparadoxes.watertracker.ApplicationScope;
import com.princeparadoxes.watertracker.data.db.model.DbDrink;
import com.princeparadoxes.watertracker.data.db.repository.DBDrinkRepository;
import com.princeparadoxes.watertracker.data.model.Drink;
import com.princeparadoxes.watertracker.data.model.StatisticType;
import com.princeparadoxes.watertracker.data.sp.ProjectPreferences;
import com.princeparadoxes.watertracker.ui.screen.main.statistic.StatisticModel;

import javax.inject.Inject;

import io.reactivex.Observable;

@ApplicationScope
public class DrinkRepository {

    private final DBDrinkRepository mDBDrinkRepository;
    private final ProjectPreferences mProjectPreferences;

    @Inject
    public DrinkRepository(DBDrinkRepository dbDrinkRepository, ProjectPreferences projectPreferences) {
        mDBDrinkRepository = dbDrinkRepository;
        mProjectPreferences = projectPreferences;

    }

    public Observable<Drink> add(Drink drink) {
        return Observable.just(DbDrink.mapFromDrink(drink))
                .map(mDBDrinkRepository::add)
                .map(Drink::mapFromDBDrink);
    }

    public Observable<Float> getDaySum() {
        return Observable.just(mDBDrinkRepository.getDayStatistic());
    }

    public Observable<Float> getLastByDay() {
        return Observable.fromCallable(mDBDrinkRepository::getLastByDay);
    }

    public Observable<StatisticModel> getSumByPeriod(StatisticType statisticType) {
        return Observable.just(mDBDrinkRepository.getAllByPeriod(statisticType))
                .map(result -> {
                    int normValue = statisticType.getCountDays() * mProjectPreferences.getCurrentDayNorm();
                    return new StatisticModel(statisticType, result, normValue);
                });
    }

    public Observable<Boolean> del(long timestamp) {
        return Observable.just(mDBDrinkRepository.deleteAllWithTimestamp(timestamp));
    }
}

File: app/src/main/java/com/princeparadoxes/watertracker/data/repository/DrinkRepository.kt
package com.princeparadoxes.watertracker.data.repository

import com.princeparadoxes.watertracker.data.mapper.DrinkMapper
import com.princeparadoxes.watertracker.data.source.db.model.DrinkSchema
import com.princeparadoxes.watertracker.data.source.db.service.DrinkDatabaseService
import com.princeparadoxes.watertracker.data.source.sp.ProjectPreferences
import com.princeparadoxes.watertracker.domain.entity.Drink
import com.princeparadoxes.watertracker.domain.entity.StatisticModel
import com.princeparadoxes.watertracker.domain.entity.StatisticType
import com.princeparadoxes.watertracker.domain.interactor.DrinkInputPort
import io.reactivex.Observable

class DrinkRepository(
        private val drinkDatabaseService: DrinkDatabaseService,
        private val preferences: ProjectPreferences
) : DrinkInputPort {

    override fun getDaySum(): Observable<Int> {
        return drinkDatabaseService.getStatisticByPeriod(StatisticType.DAY)
    }

    override fun addWater(ml: Int): Observable<Drink> {
        return Observable.just(DrinkSchema(ml, System.currentTimeMillis()))
                .map({ drinkDatabaseService.add(it) })
                .map({ DrinkMapper.mapFromDrinkSchema(it) })
    }

    override fun getLast(): Observable<Drink> {
        return drinkDatabaseService.getLast()
                .map { DrinkMapper.mapFromDrinkSchema(it) }
    }

    override fun getStatisticByPeriod(statisticType: StatisticType): Observable<StatisticModel> {
        return drinkDatabaseService.getStatisticByPeriod(statisticType)
                .map { result ->
                    val normValue = statisticType.countDays * preferences.currentDayNorm
                    StatisticModel(statisticType, result.toFloat(), normValue.toFloat())
                }
    }

    override fun getDrinksByPeriod(statisticType: StatisticType): Observable<List<Drink>> {
        return drinkDatabaseService.getDrinksByPeriod(statisticType)
                .map { it.map { DrinkMapper.mapFromDrinkSchema(it) } }
    }

    override fun removeLastDrink(): Observable<Int> {
        return Observable.just(drinkDatabaseService.removeLast())
    }

    fun del(timestamp: Long): Observable<Boolean> {
        return Observable.just(drinkDatabaseService.deleteAllWithTimestamp(timestamp)!!)
    }
}
